### PR TITLE
CMR-5407 Added 4 new related URL keywords UMM-C and upgraded to version 1.12

### PR DESCRIPTION
**Pull request by eereiter**
[related url](https://media2.giphy.com/media/wOJbAELuNAHGU/giphy-tumblr.gif?cid=6104955e5c463d3c6650765a49127c3d)

### CODE REVIEW

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj [hunk 1]
```diff
@@ -160,3 +160,93 @@
   {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "GIOVANNI"} {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "GIOVANNI"}
   {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "MAP"} {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "MAP"}
   {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "WORLDVIEW"} {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "GIBS"}})
+
+(def umm-1-11-related-url-types
```
Comment (ygliuvt) [inline umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj]: Wow! This mapping is so specific.

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj [hunk 2]
```diff
@@ -160,3 +160,93 @@
...
+   {:URLContentType "PublicationURL" :Type "VIEW RELATED INFORMATION" :Subtype "SCIENCE DATA PRODUCT VALIDATION"}
+   {:URLContentType "PublicationURL" :Type "VIEW RELATED INFORMATION" :Subtype "USER FEEDBACK PAGE"}
+   {:URLContentType "PublicationURL" :Type "VIEW RELATED INFORMATION" :Subtype "USER'S GUIDE"}
+   {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION"}
+   {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "GIOVANNI"}
+   {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "MAP"}
+   {:URLContentType "VisualizationURL" :Type "GET RELATED VISUALIZATION" :Subtype "WORLDVIEW"}])
+
+(def umm-1-12-umm-url-types->umm-1-11-umm-url-types
+  "Mapping from the UMM-C v1.11 URLContentType, Type, and Subtype to the UMM v 1.10 URLContentType, Type, and Subtype
```
Comment (ygliuvt) [inline umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj]: Update the docstring to say v1.12 and v1.11.

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj [hunk 3]
```diff
@@ -369,3 +369,28 @@
...
     collection
     (update-in-each collection [:RelatedUrls] replace-existing-related-url-keywords "1.10")))
+
+(defn replace-existing-related-url-keywords-for-umm-12
+  "Pass in a related url. If the keywords match what is in the umm-1-11-umm-url-types map then just
+   use the passed in keys because they haven't changed. If they don't match then use the UMM-12 to
+   UMM-11 mapping table."
+  [related-url]
+  (let [{:keys [URLContentType Type Subtype]} related-url
+        keywords (remove-nil-keys {:URLContentType URLContentType
```
Comment (ygliuvt) [inline umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj]: `url-types` seems to be a better name than `keywords`.

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj [hunk 4]
```diff
@@ -369,3 +369,28 @@
...
+                                   :Type Type
+                                   :Subtype Subtype})
+        result (if (nil? (some #{keywords} ru-maps/umm-1-11-related-url-types))
```
Comment (ygliuvt) [inline umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj]: Change this to 
```
result (if (some #{url-types} ru-maps/umm-1-11-related-url-types)
                 url-types
                 (ru-maps/umm-1-12-umm-url-types->umm-1-11-umm-url-types url-types))]
```

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj [hunk 5]
```diff
@@ -369,3 +369,28 @@
...
+    (assoc related-url :URLContentType (:URLContentType result)
+                       :Type (:Type result)
+                       :Subtype (:Subtype result))))
+
+(defn migrate-down-from-1_12
+  ":RelatedUrl Type and Subtypes have been added in UMM-C version 1.12 to correspond
+   to GCMD Related URL Keyword updates 8.6 December 2018 additions, so we need to translate the
+   values back to the old 8.6 keyword version."
+  [collection]
+  (if (nil? (:RelatedUrls collection))
```
Comment (ygliuvt) [inline umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj]: remove `nil?` check and flip the next two lines.

**Comment by ygliuvt** [comment]
Nice work! Thanks for doing this!

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj (hunk 1 above)
Comment (eereiter) [reply umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj]: yes it is

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj (hunk 2 above)
Comment (eereiter) [reply umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration_maps.clj]: done

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj (hunk 3 above)
Comment (eereiter) [reply umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj]: fixed

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj (hunk 4 above)
Comment (eereiter) [reply umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj]: Fixed

File: umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj (hunk 5 above)
Comment (eereiter) [reply umm-spec-lib/src/cmr/umm_spec/migration/related_url_migration.clj]: Fixed

**Pull request merged**